compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1265,24 +1266,62 @@ public void visit(BLangAssignment assignNode) {
     }
 
     public void visit(BLangRecordDestructure stmt) {
+        List<BLangExpression> varRefs = stmt.varRef.recordRefFields.stream()
+                .map(e -> e.variableReference).collect(Collectors.toList());
+        varRefs.add((BLangVariableReference) stmt.varRef.restParam);
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
     public void visit(BLangErrorDestructure stmt) {
+        List<BLangExpression> varRefs = new ArrayList<>();
+        varRefs.add(stmt.varRef.reason);
+        varRefs.addAll(stmt.varRef.detail.stream().map(e -> e.expr).collect(Collectors.toList()));
+        varRefs.add(stmt.varRef.restVar);
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
     @Override
     public void visit(BLangTupleDestructure stmt) {
+        List<BLangExpression> varRefs = new ArrayList<>(stmt.varRef.expressions);
+        varRefs.add((BLangVariableReference) stmt.varRef.restParam);
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
+    private void checkDuplicateVarRefs(List<BLangExpression> varRefs) {
+        Set<BSymbol> symbols = new HashSet<>();
+        for (BLangExpression varRef : varRefs) {
+            if (varRef == null || (varRef.getKind() != NodeKind.SIMPLE_VARIABLE_REF
```

Comment:
This logic breaks, If I have nested binding patterns. i.e `{ message : x , data : [ x, y] } = someRecord();`

Reply:
fixed in f6dd911